core/corehttp/gateway_handler_unixfs_dir.go

```diff
@@ -133,20 +139,92 @@ func (i *gatewayHandler) serveDirectory(ctx context.Context, w http.ResponseWrit
 			ShortHash: shortHash(hash),
 		}
 		dirListing = append(dirListing, di)
+		entryRequests = append(entryRequests, &entryRequest{
+			entryIndex: len(dirListing) - 1,
+			cid:       link.Cid,
+		})
+	}
+
+	// Optimization 2:
+	// Fetch in parallel entry's metadata up to FastDirIndexThreshold.
+	// Code inspired from go-merkledag.Walk().
+	// FIXME: Review number and (maybe) export it.
+	const defaultConcurrentFetch = 32
+	fetchersCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	feed := make(chan *entryRequest)
+	go func() {
+		for _, r := range entryRequests {
+			select {
+			case feed <- r:
+			case <-fetchersCtx.Done():
+				return
+			}
+		}
+	}()
+	type entryInfo struct {
+		entryIndex int
+		size string
+	}
+	out := make(chan entryInfo)
+	errChan := make(chan error)
+
+	for c := 0; c < defaultConcurrentFetch; c++ {
+		go func() {
+			for r := range feed {
+				nd, err := i.api.Dag().Get(fetchersCtx, r.cid)
+				if err != nil {
+					select {
+					case errChan <- err:
+					case <-fetchersCtx.Done():
+					}
+					return
+				}
+				file, err := unixfile.NewUnixfsFile(fetchersCtx, i.api.Dag(), nd)
+				if err != nil {
+					select {
+					case errChan <- err:
+					case <-fetchersCtx.Done():
+					}
+					return
+				}
+				size := "?"
+				if s, err := file.Size(); err == nil {
+					// Size may not be defined/supported. Continue anyways.
+					// FIXME: Check above. The UnixFS files we're iterating
+					//  (because we use the UnixFS API) should always support
+					//  this.
```

Comment:
iirc we need to keep this check because we've  seen dag-pb dir that links to dag-cbor or dag-json CIDv1